File: C++/40-数组中只出现一次的数字.cpp
class Solution
{
public:
    void FindNumsAppearOnce(vector<int> data, int *num1, int *num2)
    {
        int xor_res = 0;
        for (int i = 0; i < data.size(); i++)
        {
            xor_res ^= data[i];
        }
        int lowest_1_index_of_xor = 0;
        while ((xor_res & 1) == 0)
        {
            ++lowest_1_index_of_xor;
            xor_res = xor_res >> 1;
        }
        for (int i = 0; i < data.size(); i++)
        {
            if (!compare_onebit(data[i], lowest_1_index_of_xor))
                *num1 ^= data[i];
            else
                *num2 ^= data[i];
        }
    }
    bool compare_onebit(int num, int index)
    {
        num = num >> index;
        return num & 1;
    }
};

File: Python/40-数组中只出现一次的数字.py
# -*- coding:utf-8 -*-
class Solution:
    # 返回[a,b] 其中ab是出现一次的两个数字
    def FindNumsAppearOnce(self, array):
        # write code here
        xor = 0
        for i in range(len(array)):
            xor ^= array[i]
        reverse_bin = bin(xor)[2:][::-1]
        lowest_1_index_of_xor = reverse_bin.find('1')
        list0 = filter(lambda x: bin(x)[2:].zfill(len(reverse_bin))[::-1][lowest_1_index_of_xor] == '0', array)
        list1 = filter(lambda x: bin(x)[2:].zfill(len(reverse_bin))[::-1][lowest_1_index_of_xor] == '1', array)
        num1, num2 = 0, 0
        for t in list0:
            num1 ^= t
        for t in list1:
            num2 ^= t
        return [num1, num2]